### PR TITLE
CCZ integrity check for locale ids

### PR DESCRIPTION
**Pull request by orangejenny**
https://dimagi-dev.atlassian.net/browse/ICDS-507

Open for review. Planning to do some more testing.

### CODE REVIEW

File: corehq/apps/hqmedia/tasks.py [hunk 1]
```diff
@@ -214,6 +219,26 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
     DownloadBase.set_progress(build_application_zip, 100, 100)
 
 
+def _check_ccz_locale_id_integrity(file_cache):
+    for path in ('default/app_strings.txt', 'suite.xml'):
```
Comment (mkangia) [inline corehq/apps/hqmedia/tasks.py]: nit: `file_path` sounds better than `path`?

File: corehq/apps/hqmedia/tasks.py [hunk 2]
```diff
@@ -214,6 +219,26 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
...
+        if path not in file_cache:
+            return [_("Could not find {path} in CCZ").format(path)]
+
+    app_strings_ids = {
+        line.decode("utf-8").split('=')[0]
```
Comment (mkangia) [inline corehq/apps/hqmedia/tasks.py]: its difficult to verify this without knowing what a `line` can be?
i can see someone unaware of how app_strings are put in the ccz being confused about this.
Would it be worth to add an example here as a comment?

File: corehq/apps/hqmedia/tasks.py [hunk 3]
```diff
@@ -214,6 +219,26 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
...
+        for line in file_cache['default/app_strings.txt'].splitlines()
+    }
+
+    from corehq.apps.app_manager.xform import parse_xml
+    parsed = parse_xml(file_cache['suite.xml'])
+    suite_ids = {locale.get("id") for locale in parsed.iter("locale")}
+
+    return [
+        _("Locale ID {id} present in suite.xml but not in default app strings.").format(id=id)
+        for id in (suite_ids - app_strings_ids) if id
```
Comment (mkangia) [inline corehq/apps/hqmedia/tasks.py]: Can we add a line under method def saying something like
```
all string ids present in suite xml should be present in app_strings.
```

Why need the `if` here?

Also can the method be renamed to `find_missing_locale_ids` or something better that tells clearly about the intend of the method?

**Comment by orangejenny** [comment]
@mkangia Thanks for feedback, updates made in f688698a78c

File: corehq/apps/hqmedia/tasks.py (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/hqmedia/tasks.py]: A `None` showed up while testing; I'm guessing it's from a blank line at the end of the file.

File: corehq/apps/hqmedia/tasks.py [hunk 4]
```diff
@@ -227,3 +217,51 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
         )
 
     DownloadBase.set_progress(build_application_zip, 100, 100)
+
+
+def _find_missing_locale_ids_in_ccz(file_cache):
+    for file_path in ('default/app_strings.txt', 'suite.xml'):
+        if file_path not in file_cache:
+            return [_("Could not find {file_path} in CCZ").format(file_path)]
```
Comment (emord) [inline corehq/apps/hqmedia/tasks.py]: This for loop would only return one error at a time, but all other code paths return all errors at once. Is it worth chaging this to allow 2 errors if neither file is found?

Comment (orangejenny) [reply corehq/apps/hqmedia/tasks.py]: Pretty unlikely for even one of these to happen, but yeah, good best practice. 05d0fe04cde

**Comment by orangejenny** [comment]
Added a few tests in 08c9767 and a94317d, for the new locale id check and the old multimedia check.

File: corehq/apps/app_manager/tests/test_ccz.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
+# coding: utf-8
```
Comment (emord) [inline corehq/apps/app_manager/tests/test_ccz.py]: TIL how python decides on encoding: https://docs.python.org/2/reference/lexical_analysis.html#encoding-declarations

I was going to say we should consolidate on the recommended approach, but I think for python 3 all files are assumed to be utf-8 so its won't be required

File: corehq/apps/hqmedia/tasks.py [hunk 6]
```diff
@@ -227,3 +217,53 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
...
+def check_ccz_multimedia_integrity(domain, fpath):
+    if not toggles.CAUTIOUS_MULTIMEDIA.enabled(domain):
+        return []
+
+    errors = []
+
+    with open(fpath, 'rb') as tmp:
+        with zipfile.ZipFile(tmp, "r") as z:
+            media_suites = [f for f in z.namelist() if re.search(r'\bmedia_suite.xml\b', f)]
+            if len(media_suites) != 1:
```
Comment (emord) [inline corehq/apps/hqmedia/tasks.py]: I think this is impossible :fist_raised: :evergreen_tree:  but this doesn't handle more than one media suite

**Review by emord — APPROVED** [review]

File: corehq/apps/hqmedia/tasks.py (hunk 6 above)
Comment (orangejenny) [reply corehq/apps/hqmedia/tasks.py]: Will update if we ever support multiple media suites.

**Comment by orangejenny** [comment]
Tested on staging. @emord thank you for getting the errors displaying again.

Screen Shot 2019-05-13 at 4 52 30 PM

**Pull request merged**